### PR TITLE
Faster R-CNN object detector for the YCB dataset using torch

### PR DESCRIPTION
**Pull request by alex-mitrevski**
## Summary

This pull request adds code for training an object detector (in particular, Faster R-CNN) using PyTorch. The model we use is pretrained on the COCO data set, so we only finetune the class predictor. The functionality has been tested on the [YCB data set](http://www.ycbbenchmarks.com).

The training code is adapted from the [PyTorch object detection finetuning tutorial](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html). A script for training a model as well as a script for testing the inference have been added to `scripts` (`train_object_detector_and_classifier.py` and `test_object_detector_and_classifier.py` respectively). In addition, the training and inference workflow is described in a Jupyter notebook (`notebooks/faster_rcnn_torch.ipynb`).

Some required helper scripts from the tutorial are added as well. Sanity checking code for plotting detections has been taken from https://www.learnopencv.com/faster-r-cnn-object-detection-with-pytorch/.

## TODOs before merging

* [x] ~~The current version of the code can already train the detector on an arbitrary number of classes (at most three have been evaluated so far), but we still need to add data augmentation~~ As described in the Jupyter notebook, training is indeed done on an augmented set (although one has to be careful not to over-augment the data; otherwise, the detector is unable to detect anything on real images).
* [x] ~~It may be useful to define a generic YCB data set interface (like the COCO interface); if that is done, the training code should subsequently be updated to use that interface~~ Two dataset interfaces as required by the PyTorch DataLoader have been added (supporting either PASCAL VOC annotations or custom YAML-based annotations)
* [x] ~~We potentially need to reduce the dependency on the helper scripts from the aforementioned tutorial~~ The helper scripts are still used during training, but mostly due to the fact that they include utilities for monitoring the training loss.
* [x] The documentation has to be updated to describe how to train models and use the models for inference.

cc @minhnh @ErickKramer

### CODE REVIEW

File: augmentation/image_augmentation.py [hunk 1]
```diff
@@ -7,45 +8,49 @@
...
-    orig_coords_norm = segmented_box.segmented_coords_homog_norm[:, :2]
     # translate object coordinates to the object center's frame, i.e. whitens
-    whitened_coords_norm = orig_coords_norm - (segmented_box.x_center_norm, segmented_box.y_center_norm)
+    whitened_coords_norm = segmented_box.segmented_coords_norm - (segmented_box.x_center_norm, segmented_box.y_center_norm)
 
     # then generate a random rotation around the z-axis (perpendicular to the image plane), and limit the object scale
     # to maximum (default) 50% of the background image, i.e. the normalized largest dimension of the object must be at
     # most 0.5. To put it simply, scale the objects down if they're too big.
     # TODO(minhnh) add shear
-    max_scale = 1.
```
Comment (minhnh) [inline augmentation/image_augmentation.py]: `max_scale` greater than 1 here may lead to unexpected behaviors.

File: augmentation/image_augmentation.py [hunk 2]
```diff
@@ -7,45 +8,49 @@
...
+    max_scale = 1.75
     if segmented_box.max_dimension_norm > max_obj_size_in_bg:
         max_scale = max_obj_size_in_bg / segmented_box.max_dimension_norm
     random_rot_angle = np.random.uniform(0, np.pi)
-    rand_scale = np.random.uniform(0.5*max_scale, max_scale)
+    rand_scale = np.random.uniform(0.5, max_scale)
```
Comment (minhnh) [inline augmentation/image_augmentation.py]: `0.5 * max_scale` was correct

File: augmentation/image_augmentation.py [hunk 3]
```diff
@@ -105,56 +113,69 @@ def view_segmented_color_img(self):
...
-            mask_path = get_image_mask_path(img_path, obj_mask_dir)
-            if not os.path.exists(mask_path):
-                TerminalColors.formatted_print("skipping image '{}': mask '{}' does not exist"
-                                               .format(img_path, mask_path), TerminalColors.WARNING)
-                continue
+        # load backgrounds for image augmentation
+        self._background_paths = glob_extensions_in_directory(background_dir, ALLOWED_IMAGE_EXTENSIONS)
+        TerminalColors.formatted_print('Found {} background images '.format(len(self._background_paths)),
+                                        TerminalColors.OKBLUE)
+        # self._augment_backgrounds = []
```
Comment (minhnh) [inline augmentation/image_augmentation.py]: commented out code

File: augmentation/image_augmentation.py [hunk 4]
```diff
@@ -105,56 +113,69 @@ def view_segmented_color_img(self):
...
-            # add SegmentedObject instance for image-mask pair
             try:
-                self._segmented_objects.append(SegmentedObject(img_path, mask_path))
-            except Exception as e:
-                TerminalColors.formatted_print("failed to process image '{}' and mask '{}': {}"
-                                               .format(img_path, mask_path, e), TerminalColors.FAIL)
+                TerminalColors.formatted_print("loading images and masks for class '{}'".format(cls_name),
+                                               TerminalColors.BOLD)
+                # segmented_obj = SegmentedObjectCollection(cls_id, obj_img_dir, obj_mask_dir)
+                # self._segmented_object_collections[cls_id] = segmented_obj
```
Comment (minhnh) [inline augmentation/image_augmentation.py]: commented out code

File: augmentation/image_augmentation.py [hunk 5]
```diff
@@ -166,152 +187,194 @@ def project_segmentation_on_background(self, background_image):
...
                                        .format(split_name, annotation_path), TerminalColors.BOLD)
         if os.path.isfile(annotation_path):
             if not prompt_for_yes_or_no("file '{}' exists. Overwrite?".format(annotation_path)):
                 raise RuntimeError("not overwriting '{}'".format(annotation_path))
 
-        # store a reasonable value for the maximum number of objects projected onto each background
-        if max_obj_num_per_bg <= 0 or max_obj_num_per_bg > len(self.class_dict):
-            max_obj_num_per_bg = len(self.class_dict)
+        # # store a reasonable value for the maximum number of objects projected onto each background
+        # if max_obj_num_per_bg <= 0 or max_obj_num_per_bg > len(self.class_dict):
```
Comment (minhnh) [inline augmentation/image_augmentation.py]: commented out code

**Comment by alex-mitrevski** [comment]
> also does it still work with the tensorflow stuffs? If these are breaking changes, that should be documented.

What would not work with the tensorflow stuff? The code that's added here has nothing to do with tensorflow, as it's in a new and independent package.

> should eventually move augmentation params to a config file probably

That would be a good idea.

@ErickKramer said he'll take a look during the week.

**Comment by alex-mitrevski** [comment]
Just to add, it might be worth looking at the following repository for augmentation: https://github.com/debidatta/syndata-generation

The images generated there are meant to be more realistic, and thus likely better for learning.

**Comment by minhnh** [comment]
> tensorflow stuff? The code that's added here has nothing to do with tensorflow, as it's in a new and independent

I guess that's true. Only thing that may be related is `scripts/create_robocup_tf_record.py`, but it only use stuffs in `utils.py` I think.
